// nuby_factory/tests/com_test/components/com_maker/controllers/bar/actions/ray/ray_action.js
module.exports = {

    on_validate: function (rs){
        var self = this;this.on_input(rs);
        },

    on_input: function (rs){
        var self = this;self.on_process(rs,rs.req_props);
        },

    on_process: function (rs){
        var self = this;self.on_output(rs,input);
        },

/* ****** GET ****** */

    on_get_validate: function (rs){
        var self = this;this.on_input(rs);
        },

    on_get_input: function (rs){
        var self = this;self.on_process(rs,rs.req_props);
        },

    on_get_process: function (rs){
        var self = this;self.on_output(rs,input);
        },

/* ****** POST ****** */

    on_post_validate: function (rs){
        var self = this;this.on_input(rs);
        },

    on_post_input: function (rs){
        var self = this;self.on_process(rs,rs.req_props);
        },

    on_post_process: function (rs){
        var self = this;self.on_output(rs,input);
        },

/* ****** PUT ****** */

    on_put_validate: function (rs){
        var self = this;this.on_input(rs);
        },

    on_put_input: function (rs){
        var self = this;self.on_process(rs,rs.req_props);
        },

    on_put_process: function (rs){
        var self = this;self.on_output(rs,input);
        },

/* ****** DELETE ****** */

    on_delete_validate: function (rs){
        var self = this;this.on_input(rs);
        },

    on_delete_input: function (rs){
        var self = this;self.on_process(rs,rs.req_props);
        },

    on_delete_process: function (rs){
        var self = this;self.on_output(rs,input);
        },

    a:'a' // last comma
}
